internal/command/action.go

```diff
@@ -183,6 +194,32 @@ func joinEnv(env []string) string {
 	return strings.Join(env, "\n") + "\n"
 }
 
+// walkFn recursively searches for file starting at path and in the directories above path until it
+// is found or the root of the file system is reached. If found, returns the path to the file.
+func walkFn(file string, path string) (string, error) {
+	if filepath.IsAbs(file) {
+		return "", fmt.Errorf("file specified (%s) is an absolute path: will not recurse up", file)
+	}
+	for {
+		joinedPath := filepath.Join(path, file)
+		if _, err := os.Stat(joinedPath); err == nil {
+			// File found -- return the current filepath
+			return joinedPath, nil
+		} else if os.IsNotExist(err) {
+			upOne := filepath.Dir(path)
+			if upOne == path {
+				return "", fmt.Errorf(
+					"unable to locate file specified (%s): reached root of file system", file)
+			}
+			// Move up to parent dir
+			path = upOne
+		} else {
+			// Any other error
+			return "", fmt.Errorf("unable to locate file specified (%s): %s", file, err)
+		}
```

Comment:
One of this `else` statements can be removed as the final default case to reduce nesting:
```
if <blah> {
  something
  return foo
} else if <blah2> {
  something else
  return bar
}

return baz
```

Reply:
The code still seems a bit on the less-readable side with the ternary if/else. What about (I didn't test this though):
```
    for {
        joinedPath := filepath.Join(leafDir, secretsFile)

        _, err := os.Stat(joinedPath)

        if err != nil {
            // If the file is not present, we just move up one level and run the next loop iteration
            if os.IsNotExist(err) {
                upOne := filepath.Dir(leafDir)
                if upOne == leafDir {
                    return "", fmt.Errorf(
                        "unable to locate file specified (%s): reached root of file system", secretsFile)
                }

                leafDir = upOne
                continue
            }

            // If we have an unexpected error, we fail-fast
            return "", fmt.Errorf("unable to locate file specified (%s): %s", secretsFile, err)
        }

        // If there's no error, we found the file so we return it
        return joinedPath, nil 
    }
}
```

I'm guessing that there may be another option by using `case` too if you feel like this isn't clean either.

Reply:
I was a bit hesitant about replacing the code with your proposal because the logic is delicate and I could not reach 100% code coverage in the tests.
I thought a bit and I managed to reach 100% code coverage, which made me confident.
Then I took your code verbatim; it works fine as-is!

From a readability point of view, I am not sure which one is better, the 3 approaches are still hard to read. I am fine with any.